### PR TITLE
[Feature-13962][Plugin] Teradata data connector

### LINKED ISSUE #13962: [Feature][Plugin] Teradata data connector
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Add a data source plugin for Teradata Vantage.

Teradata publishes a JDBC driver (https://central.sonatype.com/artifact/com.teradata.jdbc/terajdbc/20.00.00.06/versions) for its flagship analytic data platform "Vantage". 

The requires creating a new plugin, with a dependency on the above "terajdbc" driver. The rest of the plugin is otherwise similar to other SQL data source plugins (eg. postgresql), so the solution is to mainly to replicate the plugin and adjust the JDBC connection string builds in the *DataSourceProcessor.java and set the default port number (1025) in the UI (use-form.ts).

### Use case

A new "Teradata" data source type s available in the data source manager. The workflow SQL step can now connect Teradata Vantage and execute SQL statements.
DolphinScheduler can be used to orchestrate workflows spanning on the Teradata platform.

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by remi-td**
## Purpose of the pull request
This pull request adds data connector plugin for Teradata Vantage analytic database.
(This continues PR #13964 that I closed by accident, sorry. Conflicts are resolved)

close https://github.com/apache/dolphinscheduler/issues/13962

## Brief change log
- Added dolphinscheduler-datasource-teradata in dolphinscheduler-datasource-plugin and as a dependency in dolphinscheduler-datasource-plugin/pom.xml
- Added Maven dependency on Teradata jdbc (v20.00.00.06) driver in dolphinscheduler-datasource-teradata/pom.xml
- Added user_doc guides for Teradata data source.
- Added Teradata as data source type in the ui
- Added Teradata driver class, JDBC and validation query to the constants.
- Added Teradata to the source types

## Verify this pull request
This pull request is code cleanup without any test coverage.

### CODE REVIEW

File: docs/docs/zh/guide/datasource/teradata.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# TERADATA 数据源
```
Comment (zhongjiajie) [inline docs/docs/zh/guide/datasource/teradata.md:1]: Did you translate the Chinese doc from a translator like google translate? :laughing:

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-teradata/pom.xml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        <dependency>
+            <groupId>org.apache.dolphinscheduler</groupId>
+            <artifactId>dolphinscheduler-datasource-api</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.teradata.jdbc</groupId>
+            <artifactId>terajdbc</artifactId>
+            <version>20.00.00.06</version>
```
Comment (zhongjiajie) [inline dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-teradata/pom.xml:47]: Dependence on `dolphinscheuler-bom`, will use this version by default
```suggestion
```

File: dolphinscheduler-ui/pnpm-lock.yaml [hunk 3]
```diff
@@ -1,21 +1,4 @@
...
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-#
-
-lockfileVersion: 5.4
+lockfileVersion: 5.3
```
Comment (zhongjiajie) [inline dolphinscheduler-ui/pnpm-lock.yaml]: it seems you downgrade the version and remove the license header. An accident?

File: dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-datasource.ts [hunk 4]
```diff
@@ -118,6 +118,11 @@ export function useDatasource(
       code: 'SSH',
       disabled: true
     },
+    {
+      id: 15,
```
Comment (zhongjiajie) [inline dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-datasource.ts]: should be `16` am I right? @devosend

File: docs/docs/zh/guide/datasource/teradata.md (hunk 1 above)
Comment (remi-td) [reply docs/docs/zh/guide/datasource/teradata.md:1]: No, just copy/paste... :) Is this wrong?

File: dolphinscheduler-ui/pnpm-lock.yaml (hunk 3 above)
Comment (remi-td) [reply dolphinscheduler-ui/pnpm-lock.yaml]: Yes, will update

File: dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-datasource.ts (hunk 4 above)
Comment (remi-td) [reply dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-datasource.ts]: yes

File: dolphinscheduler-ui/pnpm-lock.yaml (hunk 3 above)
Comment (remi-td) [reply dolphinscheduler-ui/pnpm-lock.yaml]: This is now fixed.

File: dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-datasource.ts (hunk 4 above)
Comment (remi-td) [reply dolphinscheduler-ui/src/views/projects/task/components/node/fields/use-datasource.ts]: This is fixed - commit 61cce5203b05cb4a5d2586ac4339cf723871e210

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-teradata/src/main/java/org/apache/dolphinscheduler/plugin/datasource/teradata/param/TeradataDataSourceProcessor.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        teradataDatasourceParamDTO.setDatabase(connectionParams.getDatabase());
+        teradataDatasourceParamDTO.setUserName(connectionParams.getUser());
+        teradataDatasourceParamDTO.setOther(connectionParams.getOther());
+
+        String address = connectionParams.getAddress();
+        String[] hostSeperator = address.split(Constants.DOUBLE_SLASH);
+        String[] hostPortArray = hostSeperator[hostSeperator.length - 1].split(Constants.SLASH);
+        teradataDatasourceParamDTO.setHost(hostPortArray[0].split(Constants.COLON)[0]);
+        teradataDatasourceParamDTO.setPort(
+                Integer.parseInt(connectionParams.getJdbcUrl().split("DBS_PORT=")[1].split(Constants.COMMA)[0]));
```
[the remark this answers is not in the document]
Comment (remi-td) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-teradata/src/main/java/org/apache/dolphinscheduler/plugin/datasource/teradata/param/TeradataDataSourceProcessor.java]: This is fixed - commit 61cce5203b05cb4a5d2586ac4339cf723871e210

File: dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-teradata/pom.xml (hunk 2 above)
Comment (remi-td) [reply dolphinscheduler-datasource-plugin/dolphinscheduler-datasource-teradata/pom.xml:47]: I had to roll this back: this package requires a defined version number. The JDBC drivers are changing quite slowly, this should be forward compatible for many many years... ;)

**Comment by remi-td** [comment]
Hi do you have any recommendation on how to move this one forward? The Dependency check fails on "CVSS score greater than 7" but it doesn't seem to relate to any package introduced in this branch.... Thanks

File: docs/docs/zh/guide/datasource/teradata.md (hunk 1 above)
Comment (zhongjiajie) [reply docs/docs/zh/guide/datasource/teradata.md:1]: no, that great 👍

**Comment by zhongjiajie** [comment]
It seems you missing to add jdbc license https://github.com/apache/dolphinscheduler/actions/runs/4847977685/jobs/8706488762?pr=13988

you can see example in https://github.com/apache/dolphinscheduler/commit/db6c061c8b2ac663e8b75e737937a4a342308159#diff-041065969f2c99cd524d20a995740362efa459bb3a1235505c84b7dc4c355e46
and https://github.com/apache/dolphinscheduler/commit/db6c061c8b2ac663e8b75e737937a4a342308159#diff-cbcb6920ce6ce8790f6a31eac34c50ee8f2aa8f2d3e0d2ffaa72feb7356c74a5

**Comment by zhongjiajie** [comment]
BTW, front end conflict for `dolphinscheduler-ui/pnpm-lock.yaml`

**Comment by zhongjiajie** [comment]
@RemiTurpaud It seems we have some conflict file, could you solve the conflict?

**Comment by remi-td** [comment]
I need Teradata to advise on the licensing strategy for the JDBC driver. I have to pause this feature for now and resume when I get guidance on how to proceed.
@zhongjiajie Is it OK to set it to draft? Or shall I close it and create a new one later?

**Comment by zhongjiajie** [comment]
> I need Teradata to advise on the licensing strategy for the JDBC driver. I have to pause this feature for now and resume when I get guidance on how to proceed.
> @zhongjiajie Is it OK to set it to draft? Or shall I close it and create a new one later?

draft is ok, and looking forward your come back with the permission

**Comment by zhongjiajie** [comment]
BTW, you can ping me if you have any further process

**Comment by zhongjiajie** [comment]
Hi @RemiTurpaud are we still WIP with license required process?

**Comment by zhongjiajie** [comment]
hi @RemiTurpaud Are we still license WIP?

**Comment by remi-td** [comment]
> hi @RemiTurpaud Are we still license WIP?

Hi @zhongjiajie, yes we are, and I do not have an estimate of the time it will take to compete, unfortunately...

**Review by songjianet — APPROVED** [review]

**Pull request closed without merging**